File: 2025-02-February-GFG-POTD/Get Min from Stack/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
private:
    stack<int> stack;
    stack<int> minStack;

public:
    Solution() {}

    // Add an element to the top of Stack
    void push(int x) {
        stack.push(x);
        if (minStack.empty() || x <= minStack.top()) {
            minStack.push(x);
        }
    }

    // Remove the top element from the Stack
    void pop() {
        if (!stack.empty()) {
            if (stack.top() == minStack.top()) {
                minStack.pop();
            }
            stack.pop();
        }
    }

    // Returns top element of the Stack
    int peek() {
        return stack.empty() ? -1 : stack.top();
    }

    // Finds minimum element of Stack
    int getMin() {
        return minStack.empty() ? -1 : minStack.top();
    }
};

// Driver Code
int main() {
    int T;
    cin >> T; // Number of test cases

    while (T--) {
        int q;
        cin >> q; // Number of queries
        Solution g;

        while (q--) {
            int qt;
            cin >> qt;

            if (qt == 1) {
                int att;
                cin >> att;
                g.push(att);
            } else if (qt == 2) {
                g.pop();
            } else if (qt == 3) {
                cout << g.peek() << " ";
            } else if (qt == 4) {
                cout << g.getMin() << " ";
            }
        }
        cout << "\n~\n";
    }

    return 0;
}

File: 2025-02-February-GFG-POTD/Get Min from Stack/Solution.py
class Solution:
    def __init__(self):
        self.stack = []
        self.minStack = []

    # Add an element to the top of Stack
    def push(self, x: int):
        self.stack.append(x)
        if not self.minStack or x <= self.minStack[-1]:
            self.minStack.append(x)

    # Remove the top element from the Stack
    def pop(self):
        if self.stack:
            if self.stack[-1] == self.minStack[-1]:
                self.minStack.pop()
            self.stack.pop()

    # Returns top element of the Stack
    def peek(self) -> int:
        return self.stack[-1] if self.stack else -1

    # Finds minimum element of Stack
    def getMin(self) -> int:
        return self.minStack[-1] if self.minStack else -1


# Driver Code
if __name__ == "__main__":
    T = int(input())  # Number of test cases

    for _ in range(T):
        q = int(input())  # Number of queries
        g = Solution()

        for _ in range(q):
            query = list(map(int, input().split()))
            qt = query[0]

            if qt == 1:
                g.push(query[1])
            elif qt == 2:
                g.pop()
            elif qt == 3:
                print(g.peek(), end=" ")
            elif qt == 4:
                print(g.getMin(), end=" ")

        print("\n~")
